File: Backend/src/main/java/fpt/edu/vn/Backend/service/ItemService.java
package fpt.edu.vn.Backend.service;

import fpt.edu.vn.Backend.DTO.AttachmentDTO;
import fpt.edu.vn.Backend.DTO.ItemDTO;
import fpt.edu.vn.Backend.DTO.request.AttachmentUploadDTO;
import fpt.edu.vn.Backend.DTO.request.ItemUpdateDTO;
import fpt.edu.vn.Backend.pojo.Item;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.io.IOException;
import java.util.List;

public interface ItemService {
    @NotNull Item mapDTOToEntity(@NotNull ItemUpdateDTO itemDTO, @NotNull Item item);

    @NotNull ItemDTO createItem(@NotNull ItemUpdateDTO itemDTO) throws IOException;
    @Nullable ItemDTO getItemById(int id);
    ItemDTO updateItem(@NotNull ItemUpdateDTO itemDTO);
    List<AttachmentDTO> uploadAttachment(int id, AttachmentUploadDTO dto) throws IOException;
    void deleteAttachment(int attachmentId, int itemId);

    @NotNull Page<ItemDTO> getItems(@NotNull Pageable pageable,
                                    @Nullable Integer minPrice,
                                    @Nullable Integer maxPrice,
                                    @Nullable Item.Status status,
                                    @Nullable Integer categoryId,
                                    @Nullable String search);

    @NotNull Page<ItemDTO> getItemsByPrice(@NotNull Pageable pageable, int minPrice, int maxPrice);
    @NotNull Page<ItemDTO> getItemsByStatus(@NotNull Pageable pageable, @NotNull Item.Status status);
    @NotNull Page<ItemDTO> getItemsByOwnerId(@NotNull Pageable pageable, int ownerId);
    @NotNull Page<ItemDTO> getItemsByBuyerId(@NotNull Pageable pageable, int buyerId);
    @NotNull Page<ItemDTO> getItemsByCategoryId(@NotNull Pageable pageable, int categoryId);
    @NotNull Page<ItemDTO> getItemsByCategoryId(@NotNull Pageable pageable, int categoryId, Item.Status status);
    @NotNull Page<ItemDTO> getItemsByCategoryIdByPrice(@NotNull Pageable pageable, int categoryId, int minPrice, int maxPrice);
    @NotNull Page<ItemDTO> getItemsByName(@NotNull Pageable pageable, String name);
    @NotNull Page<ItemDTO> getItemsByName(@NotNull Pageable pageable, String name,Item.Status status);
}

File: frontend/src/services/ItemService.ts
import axios from '@/config/axiosConfig.ts';
import { Item, ItemStatus } from '@/models/Item.ts';
import { Page } from '@/models/Page.ts';
import { getCookie, getBearerToken, removeCookie } from '@/utils/cookies';
import { API_SERVER } from '@/constants/domain';
import { Attachment } from '@/models/Attachment.ts';
import { ItemUpdateDTO } from '@/models/ItemUpdateDTO.ts';

// Service methods
const baseUrl = API_SERVER + '/items';

interface GetItemsSchema {
  page: number;
  size: number;
  sort?: string;
  order?: 'asc' | 'desc';
  status?: string;
  minPrice?: number;
  maxPrice?: number;
  search?: string;
  categoryId?: number;
}

export const getItems = async (input: GetItemsSchema) => {
  const { page, size, sort, status, minPrice, maxPrice, search, categoryId } = input;

  let params = {
    page: page && page - 1, // Spring Boot uses 0-based page index
    size: size ? size : 10,
    sort: sort ? sort : 'itemId,desc',
    status: status ? status.toUpperCase() : undefined,
    categoryId,
    search,
    minPrice,
    maxPrice,
  };
  return await axios.get<Page<Item>>(`${baseUrl}/`, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: getBearerToken(),
    },
    params: params,
  });
};

export const getOwnedItems = async (pageNumber: number, pageSize?: number) => {
  let params = {
    page: pageNumber,
    size: pageSize,
  };
  return await axios.get<Page<Item>>(`${baseUrl}/inventory`, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: getBearerToken(),
    },
    params: params,
  });
};

export const getItemsByCategoryId = async (
  page: number,
  size: number,
  categoryId: number,
  minPrice?: number,
  maxPrice?: number,
  sort?: string,
  order?: string,
  status?: ItemStatus
) => {
  return await axios.get<Page<Item>>(`${baseUrl}/category/${categoryId}`, {
    headers: {
      'Content-Type': 'application/json',
    },
    params: { page, size, minPrice, maxPrice, sort, order, status },
  });
};
export const getItemsByName = async (
  page: number,
  size: number,
  name: string,
  sort?: string,
  order?: string,
  status?: ItemStatus
) => {
  return await axios.get<Page<Item>>(`${baseUrl}/search/${name}`, {
    headers: {
      'Content-Type': 'application/json',
    },
    params: { page, size, sort, order, status },
  });
};

export const getItemsByStatus = async (status: ItemStatus, page: number, size: number) => {
  return await axios
    .get<Page<Item>>(`${baseUrl}/status/${status}`, {
      headers: {
        'Content-Type': 'application/json',
      },
      params: { page, size },
    });
};

export const getItemsByOwnerId = async (ownerId: number, page: number, size: number) => {
  return await axios.get<Page<Item>>(`${baseUrl}/owner/${ownerId}`, {
    headers: {
      'Content-Type': 'application/json',
    },
    params: { page, size },
  });
};

export const getItemById = async (id: number): Promise<Item | undefined> => {
  const response = await axios.get<Item>(`${baseUrl}/detail/${id}`, {
    headers: {
      'Content-Type': 'application/json',
    },
  });
  return response.data;
};

export const createItem = async (itemDTO: ItemUpdateDTO) => {
  return await axios.post<Item>(`${baseUrl}/create`, itemDTO, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: getBearerToken(),
    },
  });
};

export const updateItem = async (itemDTO: ItemUpdateDTO) => {
  if (!itemDTO.itemId) {
    throw new Error('Item id cannot be null');
  }
  return await axios.post<Item>(`${baseUrl}/update`, itemDTO, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: getBearerToken(),
    },
  });
};

export const uploadItemAttachment = async (itemId: number, data: { files: File[] }) => {
  return await axios.put<Attachment[]>(`${baseUrl}/attachment/${itemId}`, data, {
    headers: {
      'Content-Type': 'multipart/form-data',
      Authorization: getBearerToken(),
    },
  });
};

export const deleteItemAttachment = async (itemId: number, attachmentId: number) => {
  return await axios.delete<Attachment[]>(`${baseUrl}/attachment/${itemId}/${attachmentId}`, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: getBearerToken(),
    },
  });
};
